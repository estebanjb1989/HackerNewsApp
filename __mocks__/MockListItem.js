import { useNavigation } from '@react-navigation/native';
import React from 'react';
import { TouchableOpacity } from 'react-native';

export default ({ title, url }) => {
  const navigation = useNavigation()

  return (
    <TouchableOpacity 
      testID="list-item_pressable"
      onPress={() => navigation.navigate({ title, url })}
    >
      {title}
    </TouchableOpacity>
  )
}