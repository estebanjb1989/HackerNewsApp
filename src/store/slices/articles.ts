import { createSlice, createAsyncThunk, current } from "@reduxjs/toolkit";
import notifee from '@notifee/react-native';
import { IArticle, IState } from "@interfaces";

interface IArticleState {
  articles: IArticle[];
  deletedArticles: string[];
  favoriteArticles: string[];
  loading: boolean;
  query: string;
}

const initialState = {
  articles: [],
  deletedArticles: [],
  favoriteArticles: [],
  loading: false,
  query: "tech"
} as IArticleState;

const articlesSlice = createSlice({
  name: "articlesSlice",
  initialState,
  reducers: {
    setLoading(state, action) {
      state.loading = action.payload
    },
    setArticles(state, action) {
      state.articles = action.payload;
    },
    deleteArticle(state, action) {
      state.deletedArticles = [...(state.deletedArticles || []), action.payload]
    },
    addToFavorites(state, action) {
      state.favoriteArticles = [...(state.favoriteArticles || []), action.payload]
    },
    removeFromFavorites(state, action) {
      state.favoriteArticles = state.favoriteArticles?.filter(fa => fa !== action.payload)
    },
    setQuery(state, action) {
      state.query = action.payload
    }
  },
});

// Define an async action creator
export const fetchArticles = createAsyncThunk(
  'articles/fetch',
  async (args, { dispatch, getState }) => {
    try {
      dispatch(setLoading(true))
      const query = (getState() as IState).articles?.query

      const response = await fetch("https://hn.algolia.com/api/v1/search_by_date?query=" + query)
      const data = await response.json()

      const articles = data.hits
        .filter((hit: any) => hit.story_title && hit.story_url)
        .map((hit: any) => ({
          id: hit.created_at_i,
          title: hit.story_title,
          author: hit.author,
          createdAt: hit.created_at,
          url: hit.story_url
        } as IArticle))
        .sort((a: any, b: any) => {
          if (a.created_at_i > b.created_at_i) return -1;
          if (a.created_at_i < b.created_at_i) return 1;
          return 0;
        })

      const notificationsEnabled = (getState() as IState)?.settings?.notificationsEnabled
      const currentArticles = (getState() as IState)?.articles?.articles

      if (notificationsEnabled && currentArticles?.length > 0) {
        const newArticles = articles.filter((art: IArticle) =>
          !currentArticles?.some((curr: IArticle) => curr.id === art.id)
        )

        const channelId = await notifee.createChannel({
          id: 'default',
          name: 'Default Channel',
        });

        for (const newArticle of newArticles) {
          await notifee.displayNotification({
            title: "New article by " + newArticle.author,
            body: newArticle.title,
            data: {
              url: newArticle.url || ""
            },
            android: {
              channelId,
              smallIcon: 'ic_launcher', // optional, defaults to 'ic_launcher'.
              // pressAction is needed if you want the notification to open the app when pressed
              pressAction: {
                id: 'default',
              },
            },
          });
        }
      }

      return articles
    } catch (err) {
      console.error(err)
      throw err
    } finally {
      dispatch(setLoading(false))
    }
  }
);

export const {
  setLoading,
  setArticles,
  deleteArticle,
  addToFavorites,
  removeFromFavorites,
  setQuery
} = articlesSlice.actions;

export default articlesSlice.reducer;
