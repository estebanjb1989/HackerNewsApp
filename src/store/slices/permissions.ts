import { createSlice } from "@reduxjs/toolkit";

interface IPermissionsState {
  notificationsPermissionStatus: number | null;
}

const initialState = {
  notificationsPermissionStatus: null,
} as IPermissionsState;

const permissionsSlice = createSlice({
  name: "permissionsSlice",
  initialState,
  reducers: {
    setNotificationsPermissionStatus(state, action) {
      state.notificationsPermissionStatus = action.payload
    },    
  },
});

export const {
  setNotificationsPermissionStatus
} = permissionsSlice.actions;

export default permissionsSlice.reducer;
