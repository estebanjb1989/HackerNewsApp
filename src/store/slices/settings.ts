import { createSlice } from "@reduxjs/toolkit";

interface ISettingsState {
  notificationsEnabled: boolean;
}

const initialState = {
  notificationsEnabled: false,
} as ISettingsState;

const settingsSlice = createSlice({
  name: "settingsSlice",
  initialState,
  reducers: {
    setNotificationsEnabled(state, action) {
      state.notificationsEnabled = action.payload
    },    
  },
});

export const {
  setNotificationsEnabled
} = settingsSlice.actions;

export default settingsSlice.reducer;
