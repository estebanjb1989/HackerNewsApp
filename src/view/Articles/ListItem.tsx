import React from 'react';
import {
  StyleSheet,
  Text,
  View,
  Image,
  TouchableOpacity
} from 'react-native';
import { ParamListBase, useNavigation } from '@react-navigation/native';
import { DrawerNavigationProp } from "@react-navigation/drawer";
import { IArticle, IStory } from '@interfaces';
import FavoriteIcon from "@img/icons/favorite.png"

export const getTimeAgoFromTS = (timestamp: number) => {
  const now = new Date();
  const diff = now.getTime() - timestamp;
  const seconds = Math.floor(diff / 1000);
  const minutes = Math.floor(seconds / 60);
  const hours = Math.floor(minutes / 60);
  const days = Math.floor(hours / 24);

  if (days > 0) {
    return `${days} ${days === 1 ? 'day' : 'days'} ago`;
  } else if (hours > 0) {
    return `${hours} ${hours === 1 ? 'hour' : 'hours'} ago`;
  } else if (minutes > 0) {
    return `${minutes} ${minutes === 1 ? 'minute' : 'minutes'} ago`;
  } else if (seconds > 60) {
    return `${Math.floor(seconds / 60)} ${seconds % 60} minutes ago`;
  } else {
    return `${seconds} seconds ago`;
  }
}

interface IListItemProps {
  item: IArticle;
  onLongPress: (id: string) => void;
  isFavorite: boolean;
}

function ListItem({
  item,
  onLongPress,
  isFavorite
}: IListItemProps): React.JSX.Element {
  const navigation = useNavigation<DrawerNavigationProp<ParamListBase>>()
  

  const handlePress = () => navigation.navigate("Article",
    { url: item.url, title: item.title }
  )

  const timeAgo = getTimeAgoFromTS(new Date(item.createdAt).getTime())

  return (
    <TouchableOpacity
      testID="list-item_pressable"
      onPress={handlePress}
      onLongPress={() => onLongPress?.(item.id)}
    >
      <View style={styles.container}>
        <View style={styles.infoContainer}>
          <View style={styles.infoTopContainer}>
            {isFavorite ? (
              <Image source={FavoriteIcon} style={styles.favoriteIcon}></Image>
            ) : null}
          </View>
          <Text>{item.title}</Text>
          <Text style={styles.authorText} testID="list-item_author">{item.author} - {timeAgo}</Text>
        </View>
      </View>
    </TouchableOpacity>
  );
}

const styles = StyleSheet.create({
  container: {
    flexDirection: "row",
    padding: 12,
    gap: 12,
    alignItems: "center",
    borderWidth: .5,
    borderStyle: "solid",
    borderColor: "grey",
    marginHorizontal: 8,
  },
  infoTopContainer: {
    flexDirection: "row",
    justifyContent: "space-between"
  },
  infoContainer: {
    flex: 1
  },
  favoriteIcon: {
    width: 24, 
    height: 24,
    position: "absolute",
    right: -12,
    top: -12,
  },
  authorText: {
    marginTop: 8,
    color: "grey"
  }
});

export default ListItem;
