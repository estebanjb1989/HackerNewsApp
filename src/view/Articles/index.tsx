import React, { useMemo, useEffect, useCallback } from 'react';
import {
  SafeAreaView,
  FlatList,
  StyleSheet,
  View,
  TouchableOpacity,
  Text,
  TextInput,
  RefreshControl
} from 'react-native';
import { throttle } from 'lodash';
import ListItem from "./ListItem"
import _ from "lodash"
import { useDispatch, useSelector } from 'react-redux';
import { useIsConnected } from 'react-native-offline';
import { Swipeable } from 'react-native-gesture-handler';
import { IState, IArticle } from "@interfaces"
import Title from "@component/Title"
import Spacer from "@component/Spacer"
import {
  setArticles,
  deleteArticle,
  addToFavorites,
  removeFromFavorites,
  fetchArticles,
  setQuery
} from '@store/slices/articles';
import DeleteIcon from "@img/icons/delete.png"

const Separator = () => <View style={{ height: 10 }} />;

function Articles() {
  const isConnected = useIsConnected()
  const dispatch = useDispatch()
  const [
    articles,
    deletedArticles,
    favoriteArticles,
    loading,
  ] = useSelector((state: IState) => [
    state.articles.articles,
    state.articles.deletedArticles,
    state.articles.favoriteArticles,
    state.articles.loading
  ])

  const articlesMemo = useMemo(() => {
    return articles?.filter?.((article: IArticle) => 
      !deletedArticles?.some((da: string) => da === article.id))
  }, [articles, deletedArticles])

  const onRefresh = useCallback(async () => {
    if (!isConnected) return
    const data = await dispatch(fetchArticles())
    dispatch(setArticles(data?.payload))
  }, [isConnected])

  useEffect(() => {
    onRefresh()
  }, [isConnected])

  const handleItemDelete = (id: string) => () => {
    dispatch(deleteArticle(id))
  }

  const handleItemLongPress = useCallback((id: string) => () => {
    if (favoriteArticles?.some((fa: string) => fa === id)) {
      dispatch(removeFromFavorites(id))
    } else {
      dispatch(addToFavorites(id))
    }
  }, [favoriteArticles])

  const handleTextInput = (text: string) => {
    dispatch(setQuery(text))
  };

  const throttledHandleTextInput = throttle(handleTextInput, 2000);

  return (
    <SafeAreaView>
      <Title text="Home" />
      <TextInput
        testID="articles_text-input"
        onChangeText={(text) => throttledHandleTextInput(text)}
        style={styles.textInput}
        placeholder='Search by topic... (iOS, Android, etc.)'
        placeholderTextColor="grey"
        onSubmitEditing={onRefresh}
      />
      <FlatList
        testID="articles_flat-list"
        data={articlesMemo}
        keyExtractor={(item) => item.id}
        windowSize={10}
        refreshControl={
          <RefreshControl
            refreshing={loading}
            onRefresh={onRefresh}
          />
        }
        renderItem={({ item }) => (
          <Swipeable
            renderRightActions={() => (
              <TouchableOpacity onPress={handleItemDelete(item.id)}>
                <View style={styles.deleteButton}>
                  <Text style={styles.deleteButtonText}>Delete</Text>
                </View>
              </TouchableOpacity>
            )}
          >
            <ListItem
              onLongPress={handleItemLongPress(item.id)}
              key={item.title}
              item={item}
              isFavorite={favoriteArticles?.some((fa: string) => fa === item.id)}
            />
          </Swipeable>
        )}
        ItemSeparatorComponent={Separator}
        ListFooterComponent={<Spacer height={120} />}
      />

    </SafeAreaView>
  );
}

const styles = StyleSheet.create({
  textInput: {
    backgroundColor: "white",
    marginBottom: 12,
    marginHorizontal: 12,
    color: "black"
  },
  deleteButton: {
    backgroundColor: "red",
    width: 80,
    height: "100%",
    justifyContent: "center",
    alignItems: "center",
    marginLeft: -8
  },
  deleteButtonText: {
    textAlign: "center",
    width: 80,
    color: "white"
  }
});

export default Articles;
