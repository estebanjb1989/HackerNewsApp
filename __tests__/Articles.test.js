import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import { useSelector } from "react-redux"
import Articles from "@view/Articles"
import * as articlesSlice from "@store/slices/articles"
import * as offline from "react-native-offline"
import MockListItem from "../__mocks__/MockListItem"

const spySetQuery = jest.spyOn(articlesSlice, "setQuery");
const spyFetchArticles = jest.spyOn(articlesSlice, "fetchArticles")
const spySetArticles = jest.spyOn(articlesSlice, "setArticles")
const spyUseIsConnected = jest.spyOn(offline, "useIsConnected")
  .mockImplementation(() => true)

jest.mock('react-native-gesture-handler', () => ({
  Swipeable: () => <MockListItem title="title" url="www.google.com" />
}));

jest.mock("@store/slices/articles", () => ({  
  getState: () => ({
    settings: jest.fn(),
    articles: jest.fn()
  }),
  setQuery: jest.fn(),
  fetchArticles: jest.fn(),
  setArticles: jest.fn()
}))

const mockState = {
  articles: {
    articles: [{
      id: "1111111111",
      author: "author1",
    }, {
      id: "2222222222",
      author: "author2",
    }, {
      id: "3333333333",
      author: "author3",
    }, {
      id: "4444444444",
      author: "author4",
    }],
    favoriteArticles: ["3333333333"],
    deletedArticles: ["4444444444"]
  }
}

jest.mock("react-redux", () => ({
  useSelector: jest.fn(),
  useDispatch: () => jest.fn()
}))

const mockNavigate = jest.fn()

jest.mock("@react-navigation/native", () => ({
  useNavigation: () => ({
    navigate: mockNavigate
  })
}))

jest.mock("react-native-offline", () => ({
  useIsConnected: jest.fn().mockReturnValue(true)
}))

describe("Main Articles view", () => {
  beforeEach(() => {
    useSelector.mockImplementation((cb) => cb(mockState))
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  it('renders correctly', () => {
    const tree = render(<Articles />);
    expect(tree).toMatchSnapshot();
  });

  it('should render all the articles but the deleted one', async () => {
    const tree = render(<Articles />);
    const flatList = await tree.findByTestId("articles_flat-list")
    expect(flatList.props?.data?.length).toBe(3)
    expect(flatList.props?.data?.some(item => item.author === "author4")).toBeFalsy()
  });

  it("should dispatch setQuery on text change", async () => {
    const tree = render(<Articles />);
    const input = await tree.findByTestId("articles_text-input")
    fireEvent.changeText(input, 'Android');
    expect(spySetQuery).toHaveBeenCalled()
  })

  it('should dispatch fetchArticles and setArticles on submit text', async () => {
    const tree = render(<Articles />);
    const input = await tree.findByTestId("articles_text-input")
    fireEvent(input, 'submitEditing');
    expect(spyFetchArticles).toHaveBeenCalled()
    expect(spySetArticles).toHaveBeenCalled()    
  });

  it('should not dispatch fetchArticles and setArticles on submit text if device is offline', async () => {
    spyUseIsConnected.mockImplementation(() => false)
    const tree = render(<Articles />);
    const input = await tree.findByTestId("articles_text-input")
    fireEvent(input, 'submitEditing');
    expect(spyFetchArticles).not.toHaveBeenCalled()
    expect(spySetArticles).not.toHaveBeenCalled()    
  });

  it('should navigate to Article/WebView view with corresponding url if ListItem is pressed', async () => {    
    const tree = render(<Articles />);
    const listItem = await tree.queryAllByTestId("list-item_pressable")?.[0]
    fireEvent(listItem, 'press')
    expect(mockNavigate).toHaveBeenCalledWith({
      title: "title",
      url: "www.google.com"
    })
  });
})

