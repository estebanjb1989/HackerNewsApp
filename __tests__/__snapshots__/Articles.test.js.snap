// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main Articles view renders correctly 1`] = `
<RCTSafeAreaView>
  <Text
    style={
      [
        {
          "fontSize": 20,
          "marginHorizontal": 8,
          "marginVertical": 8,
          "minWidth": 240,
        },
        undefined,
      ]
    }
  >
    Home
  </Text>
  <TextInput
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Search by topic... (iOS, Android, etc.)"
    placeholderTextColor="grey"
    style={
      {
        "backgroundColor": "white",
        "color": "black",
        "marginBottom": 12,
        "marginHorizontal": 12,
      }
    }
    testID="articles_text-input"
  />
  <RCTScrollView
    ItemSeparatorComponent={[Function]}
    ListFooterComponent={
      <Spacer
        height={120}
      />
    }
    data={
      [
        {
          "author": "author1",
          "id": "1111111111",
        },
        {
          "author": "author2",
          "id": "2222222222",
        },
        {
          "author": "author3",
          "id": "3333333333",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    refreshControl={
      <RefreshControlMock
        onRefresh={[Function]}
      />
    }
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={0.0001}
    stickyHeaderIndices={[]}
    testID="articles_flat-list"
    viewabilityConfigCallbackPairs={[]}
    windowSize={10}
  >
    <RCTRefreshControl />
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
          testID="list-item_pressable"
        >
          title
        </View>
        <View
          style={
            {
              "height": 10,
            }
          }
        />
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
          testID="list-item_pressable"
        >
          title
        </View>
        <View
          style={
            {
              "height": 10,
            }
          }
        />
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
          testID="list-item_pressable"
        >
          title
        </View>
      </View>
      <View
        onLayout={[Function]}
      >
        <View
          style={
            {
              "height": 120,
            }
          }
        />
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
